File: examples/python/quickstart_tls.py
#!/usr/bin/env python3
"""ferrokinesis quickstart — boto3 over TLS

Prerequisites:
    pip install -r requirements.txt
    ferrokinesis running with TLS:
        cargo install ferrokinesis --features tls
        ferrokinesis generate-cert
        ferrokinesis --tls-cert cert.pem --tls-key key.pem

Usage:
    python quickstart_tls.py
    KINESIS_ENDPOINT=https://localhost:4567 python quickstart_tls.py
"""
import os
import time
import urllib3

import boto3

# Suppress InsecureRequestWarning for self-signed certs
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

ENDPOINT = os.environ.get("KINESIS_ENDPOINT", "https://localhost:4567")
STREAM = "python-example-tls"

client = boto3.client(
    "kinesis",
    endpoint_url=ENDPOINT,
    region_name="us-east-1",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    verify=False,
)


def wait_for_active(stream_name, max_retries=30):
    for _ in range(max_retries):
        resp = client.describe_stream(StreamName=stream_name)
        if resp["StreamDescription"]["StreamStatus"] == "ACTIVE":
            return
        time.sleep(0.2)
    raise TimeoutError(f"Stream {stream_name!r} did not become ACTIVE")


# Create a stream
print("==> CreateStream")
client.create_stream(StreamName=STREAM, ShardCount=2)
wait_for_active(STREAM)

# Put a record
print("==> PutRecord")
put = client.put_record(StreamName=STREAM, Data=b"hello world", PartitionKey="pk1")
shard_id = put["ShardId"]

# Get records
print("==> GetRecords")
iterator = client.get_shard_iterator(
    StreamName=STREAM,
    ShardId=shard_id,
    ShardIteratorType="TRIM_HORIZON",
)["ShardIterator"]

records = client.get_records(ShardIterator=iterator)
for r in records["Records"]:
    print(f"{r['PartitionKey']}: {r['Data'].decode()}")

# Clean up
print("==> DeleteStream")
client.delete_stream(StreamName=STREAM)

print("Done.")

File: examples/node/quickstart-tls.mjs
// ferrokinesis quickstart — AWS SDK for Node.js (v3) over TLS
//
// Prerequisites:
//   npm install
//   ferrokinesis running with TLS:
//     cargo install ferrokinesis --features tls
//     ferrokinesis generate-cert
//     ferrokinesis --tls-cert cert.pem --tls-key key.pem
//
// Usage:
//   node quickstart-tls.mjs
//   KINESIS_ENDPOINT=https://localhost:5000 node quickstart-tls.mjs

import {
  KinesisClient,
  CreateStreamCommand,
  DescribeStreamCommand,
  PutRecordCommand,
  GetShardIteratorCommand,
  GetRecordsCommand,
  DeleteStreamCommand,
} from "@aws-sdk/client-kinesis";
import { NodeHttpHandler } from "@smithy/node-http-handler";
import https from "https";

const ENDPOINT = process.env.KINESIS_ENDPOINT || "https://localhost:4567";
const STREAM = "node-example-tls";

const client = new KinesisClient({
  endpoint: ENDPOINT,
  region: "us-east-1",
  credentials: { accessKeyId: "test", secretAccessKey: "test" },
  requestHandler: new NodeHttpHandler({
    httpsAgent: new https.Agent({ rejectUnauthorized: false }),
  }),
});

async function waitForActive(streamName) {
  for (let i = 0; i < 30; i++) {
    const resp = await client.send(
      new DescribeStreamCommand({ StreamName: streamName })
    );
    if (resp.StreamDescription?.StreamStatus === "ACTIVE") return;
    await new Promise((r) => setTimeout(r, 200));
  }
  throw new Error(`Stream ${streamName} did not become ACTIVE`);
}

// Create a stream
console.log("==> CreateStream");
await client.send(
  new CreateStreamCommand({ StreamName: STREAM, ShardCount: 2 })
);
await waitForActive(STREAM);

// Put a record
console.log("==> PutRecord");
const put = await client.send(
  new PutRecordCommand({
    StreamName: STREAM,
    Data: new TextEncoder().encode("hello world"),
    PartitionKey: "pk1",
  })
);

// Get records
console.log("==> GetRecords");
const iterResp = await client.send(
  new GetShardIteratorCommand({
    StreamName: STREAM,
    ShardId: put.ShardId,
    ShardIteratorType: "TRIM_HORIZON",
  })
);

const records = await client.send(
  new GetRecordsCommand({ ShardIterator: iterResp.ShardIterator })
);

for (const record of records.Records) {
  const data = new TextDecoder().decode(record.Data);
  console.log(`${record.PartitionKey}: ${data}`);
}

// Clean up
console.log("==> DeleteStream");
await client.send(new DeleteStreamCommand({ StreamName: STREAM }));

console.log("Done.");
